contracts/exchange/mixins/EndemicSale.sol

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.18;
+
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+import "./EndemicFundsDistributor.sol";
+import "./EndemicExchangeCore.sol";
+import "./EndemicEIP712.sol";
+import "./EndemicNonceManager.sol";
+
+abstract contract EndemicSale is
+    ReentrancyGuardUpgradeable,
+    EndemicFundsDistributor,
+    EndemicExchangeCore,
+    EndemicEIP712,
+    EndemicNonceManager
+{
+    using ECDSA for bytes32;
+
+    bytes32 private constant SALE_TYPEHASH =
+        keccak256(
+            "Sale(uint256 orderNonce,address nftContract,uint256 tokenId,address paymentErc20TokenAddress,uint256 price,address buyer,uint256 expiresAt)"
+        );
+
+    struct Sale {
+        address seller;
+        uint256 orderNonce;
+        address nftContract;
+        uint256 tokenId;
+        address paymentErc20TokenAddress;
+        uint256 price;
+        address buyer;
+        uint256 expiresAt;
+    }
+
+    event SaleSuccess(
+        address indexed nftContract,
+        uint256 indexed tokenId,
+        address indexed seller,
+        address buyer,
+        uint256 price,
+        uint256 totalFees,
+        address paymentErc20TokenAddress
+    );
+
+    error SaleExpired();
+
+    function buyFromSale(
+        uint8 v,
+        bytes32 r,
+        bytes32 s,
+        Sale calldata sale
+    ) external payable nonReentrant {
+        if (block.timestamp > sale.expiresAt) revert SaleExpired();
+
+        if (
+            (sale.buyer != address(0) && sale.buyer != msg.sender) ||
```

Comment:
Sale ima buyera u slucaju rezervacije?

Reply:
Da, ako seller zeli rezervirat nft za nekog onda potpisuje tog buyera i onda samo taj buyer moze pozvat funkciju, i to je taj uvjet u 59. liniji. Ako seller ne zeli rezervirat onda potpisuje buyer=address(0) i onda bilo tko moze kupit nft.